packages/@aws-cdk/aws-servicediscovery/lib/non-ip-instance.ts

```diff
@@ -0,0 +1,58 @@
+import cdk = require('@aws-cdk/cdk');
+import { BaseInstanceProps, InstanceBase } from './instance';
+import { NamespaceType } from './namespace';
+import { IService } from './service';
+import { CfnInstance } from './servicediscovery.generated';
+
+// tslint:disable-next-line:no-empty-interface
+export interface NonIpInstanceBaseProps extends BaseInstanceProps {
+}
+
+/*
+ * Properties for a NonIpInstance
+ */
+export interface NonIpInstanceProps extends NonIpInstanceBaseProps {
+  /**
+   * The Cloudmap service this resource is registered to.
+   */
+  service: IService;
+}
+
+/*
+ * Instance accessible using values other than an IP address or a domain name (CNAME).
+ * Specify the other values in Custom attributes.
+ */
+export class NonIpInstance extends InstanceBase {
+  /**
+   * The Id of the instance
+   */
+  public readonly instanceId: string;
+
+  /**
+   * The Cloudmap service to which the instance is registered.
+   */
+  public readonly service: IService;
+
+  constructor(scope: cdk.Construct, id: string, props: NonIpInstanceProps) {
+    super(scope, id);
+
+    if (props.service.namespace.type !== NamespaceType.Http) {
+      throw new Error('This type of instance can only be registered for HTTP namespaces.');
```

Comment:
All of those runtime errors feel like they should be caught statically by a type checker. I'd really like it so, but I guess I also see how this could cause a lot of duplication... That's one more of those cases where support for generics in `jsii` could have been helpful.